ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,807 @@
+          for (GetChunk dataChunk : dataChunks) {
+            if (dataChunk.isFree() && chunkIdIterator.hasNext()) {
+              dataChunk.initialize(chunkIdIterator.nextIndex(), (BlobId) chunkIdIterator.next());
+            }
+            if (dataChunk.isInProgress() || (dataChunk.isReady()
+                && numChunksRetrieved - getBlobResult.getNumChunksWrittenOut() < NonBlockingRouter.MAX_IN_MEM_CHUNKS)) {
+              dataChunk.poll(requestRegistrationCallback);
+              if (dataChunk.isComplete()) {
+                onChunkOperationComplete(dataChunk);
+                if (operationCompleted) {
+                  break;
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    if (operationException.get() != null) {
+      abort(operationException.get());
+    }
+  }
+
+  // ReadableStreamChannel implementation:
+
+  /**
+   * A class that implements the result of this GetBlobOperation. This is instantiated if/when the first data chunk of
+   * the blob arrives, when the operation callback is invoked.
+   */
+  private class GetBlobResult implements ReadableStreamChannel {
+    // whether this ReadableStreamChannel is open.
+    private boolean isOpen = true;
+    // whether readInto() has been called yet by the caller on this ReadableStreamChannel.
+    private volatile boolean readCalled = false;
+    // The channel to write chunks of the blob into. This will be initialized when the caller calls the readInto().
+    private AsyncWritableChannel asyncWritableChannel;
+    // the callback to call when all the chunks are successfully written out into the asyncWritableChannel.
+    private Callback<Long> readIntoCallback;
+    // the future to mark as done when all the chunks are successfully written out into the asyncWritableChannel.
+    private FutureResult<Long> readIntoFuture;
+    // size of the data in this channel. This will be set eventually.
+    private volatile long streamSize = -1;
+    // the number of bytes written out to the asyncWritableChannel. This would be the size of the blob eventually.
+    private Long bytesWritten = 0L;
+    // the number of chunks that have been written out to the asyncWritableChannel.
+    private final AtomicInteger numChunksWrittenOut = new AtomicInteger(0);
+    // the index of the next chunk that is to be written out to the asyncWritableChannel.
+    private int indexOfNextChunkToWriteOut = 0;
+    // whether this object has called the readIntoCallback yet.
+    private AtomicBoolean readIntoCallbackCalled = new AtomicBoolean(false);
+    // the callback that is passed into the asyncWritableChannel write() operation.
+    Callback<Long> chunkAsyncWriteCallback = new Callback<Long>() {
+      @Override
+      public void onCompletion(Long result, Exception exception) {
+        bytesWritten += result;
+        if (exception != null) {
+          operationException.set(exception);
+        }
+        numChunksWrittenOut.incrementAndGet();
```

Comment:
I might have asked this before but still not sure why this needs to be an `AtomicInteger`? Should be enough if it is `volatile`, just like `bytesWritten`?